src/listpack.c

```diff
@@ -944,6 +952,25 @@ unsigned char *lpInsertInteger(unsigned char *lp, long long lval, unsigned char
     return lpInsert(lp, NULL, intenc, enclen, p, where, newp);
 }
 
+/* This is just a wrapper for lpInsert() to directly take a double.
+ * If the double happens have a value representable by long long, it'll be inserted as an integer,
+ * if not, it'll be converted to string.
+ * The advantage of exposing such an interface is that the caller doesn't convert doubles to strings,
+ * and then the listpack code tries to see if they happen to be integers (inefficient). */
+unsigned char *lpInsertDouble(unsigned char *lp, double dval, unsigned char *p, int where, unsigned char **newp) {
+    long long lval;
+    if (double2ll(dval, &lval)) {
+        uint64_t enclen; /* The length of the encoded element. */
+        unsigned char intenc[LP_MAX_INT_ENCODING_LEN];
+        lpEncodeIntegerGetType(lval, intenc, &enclen);
+        return lpInsert(lp, NULL, intenc, enclen, p, where, newp);
+    } else {
+        char sval[MAX_D2STRING_CHARS];
+        int slen = d2string(sval,sizeof(sval),dval);
+        return lpInsert(lp, (unsigned char *)sval, NOT_INT, slen, p, where, newp);
```

Comment:
I feel this is wrong, because `double2ll` determines the range of `-4503599627370495 ~ 4503599627370495`, when double val is not in this range (e.g 4503599627370496), it can still be converted to int64, but now it is converted to string.

Reply:
that's a problem in double2ll, not really related to it's use in lpInsertDouble.
or am i missing something?

i.e. every non fractional number in the above mentioned range (53 bits) can be safely converted between double and 64 bit integer without any precision loss. outside that range, some values can still be converted, and others do not.
i.e. we have 52 bits for the part, and 11 bit exponent.
so we can take any value in the above range and multiply it by 2, 4, 8 and still represent it correctly.
but since numbers in this set are sparse (unlike the ones inside the range), i think we should keep overlooking that.

assuming we're referring to the same thing, i don't think this is likely to cause any real impact to zset scores, or any future user of lpInsertDouble.

Reply:
When we execute a zadd zset 94503599627370495 m, we will do the following
1) Convert the string "94503599627370495" to double, at this point double val will be 94503599627370496 and the precision of double will be lost

```c
getDoubleFromObjectOrReply
  -> getDoubleFromObject
         -> string2d
               -> strtod
```

This step will change
2) In the old code, we will store 94503599627370496 as int64 in the listpack
2) In the new code, we will store 94503599627370496 as a string in the listpack


```
127.0.0.1:6379> zadd key 94503599627370495 m
(integer) 1
127.0.0.1:6379> zrange key 0 -1 withscores
1) "m"
2) "94503599627370496"
```

Reply:
Good explanation @sundb. But maybe it's not important? Not "likely to cause any real impact"?

Reply:
@zuiderkwast But memory consumption will increase, which is this pr is trying to solve, and cpu gains are not increasing much.

Reply:
so `strtod` loses precision and results with a double outside that range that could still be represented as an integer?
i.e. it could have produced a fractional value that's closer to the requested one.
is that right or am i still missing something?

Reply:
It could also be a double outside the safe range but that's still equal to a long. Not sure how many such numbers there are.

Reply:
Will this solve the problem?

```suggestion
        int is_huge = dval < -4503599627370495 || dval > 4503599627370495;
        return lpInsert(lp, (unsigned char *)sval, is_huge ? NULL : NOT_INT, slen, p, where, newp);
```

Reply:
> It could also be a double outside the safe range but that's still equal to a long. Not sure how many such numbers there are.

    if `double val <= LLONG_MAX && double val >= LLONG_MIN`, the following code always holds.

```c
d = strtod(...);
assert(d == (long long)d);
```

> so `strtod` loses precision and results with a double outside that range that could still be represented as an integer? i.e. it could have produced a fractional value that's closer to the requested one. is that right or am i still missing something?
    
    Yes.

```sh
127.0.0.1:6379> zadd key 94503599627370495.1 m
(integer) 0
127.0.0.1:6379> zrange key 0 -1 withscores
1) "m"
2) "94503599627370496"
```

Reply:
i see it's not only a limitation of `strtod`, but also the compiler / FPU.
so basically, there are no fractional numbers outside our range.
i.e. all numbers outside that range have no fractional part.
some are still within the 64bit integer range, and others are outside that range too, but all without fractional part.
is that right?

Reply:
@oranagra Yes.

Reply:
Wow, then just remove the interval check? (or use LLONG_MIN and LLONG_MAX instead)

Reply:
this rocks my world. i need to dig deeper to learn more about it.

Reply:
ok, managed to wrap my head around it.. it's actually really simple, don't know why i didn't realize it right away.

the exponent is 11 bit multiplier, or divisor (we can multiply by 2, 4, 8, etc, or we can divide by 2,4,8, etc), but we can't have both, and we can't multiply by a fractional part.
so the only way to get a number that's higher than 2^52 would be to multiply it by 2,4,8, etc, and this way there are no fractional digits.

i.e. my earlier statement saying "it could have produced a fractional value that's closer to the requested one" was wrong... it can't.

in that light it means that the thresholds of 2^52 could be dropped and replaced with thresholds of 2^63.
i.e. any double up to 2^63 can be represented by long long.
but not any long long up to 2^63 can be represented by double (in the high double can't represent some of them, but it has a wider range of up to 2^51*1024)

Reply:
Makes sense. If the range check doesn't improve performance compare to just a cast, we can as well drop it and just rely on the check `value == (double)(long long)value`, right?

I have no opinion about whether to add the insert-double API or not.